src/Repulsion.cc

```diff
@@ -60,35 +60,39 @@ std::shared_ptr<BHNode> BHTree::InitializeRoot(const aamcc::NucleonVector* nucle
   }
 
   G4ThreeVector cr = {0.0, 0.0, 0.0};
-  size_t nucleons_sz = nucleons->size();
+  double nucleons_sz = static_cast<double>(nucleons->size());
+  double sumX = 0.0;
+  double sumY = 0.0;
+  double sumZ = 0.0;
+
   for (const auto& nuc : *nucleons) {
-    cr.setX(cr.x() + nuc.GetX());
-    cr.setY(cr.y() + nuc.GetY());
-    cr.setZ(cr.z() + nuc.GetZ());
+    sumX += nuc.GetX();
+    sumY += nuc.GetY();
+    sumZ += nuc.GetZ();
   }
-  cr.setX(cr.x() / static_cast<double>(nucleons_sz));
-  cr.setY(cr.y() / static_cast<double>(nucleons_sz));
-  cr.setZ(cr.z() / static_cast<double>(nucleons_sz));
+  cr.setX(sumX / nucleons_sz);
+  cr.setY(sumY / nucleons_sz);
+  cr.setZ(sumZ / nucleons_sz);
 
   double maxRange = std::max({maxX - minX, maxY - minY, maxZ - minZ});
-  return std::make_shared<BHNode>(maxRange, G4ThreeVector((minX + maxX) / 2, (minY + maxY) / 2, (minZ + maxZ) / 2));
+  return std::make_unique<BHNode>(maxRange, G4ThreeVector((minX + maxX) / 2, (minY + maxY) / 2, (minZ + maxZ) / 2));
 }
 
-std::shared_ptr<BHNode> BHTree::BuildBHTree(const aamcc::NucleonVector* nucleons) {
+void BHTree::BuildBHTree(const aamcc::NucleonVector* nucleons) {
   rootnode_ = InitializeRoot(nucleons);
 
   for (size_t i = 0; i < nucleons->size(); i++) {
-    InsertNucleon(rootnode_, {(*nucleons)[i].GetX(), (*nucleons)[i].GetY(), (*nucleons)[i].GetZ()}, i);
+    G4ThreeVector vec = {(*nucleons)[i].GetX(), (*nucleons)[i].GetY(), (*nucleons)[i].GetZ()};
+    rootnode_ = InsertNucleon(std::move(rootnode_), vec, i);
   }
-  return rootnode_;
 }
 
-void BHTree::InsertNucleon(std::shared_ptr<BHNode>& node, const G4ThreeVector cords, int pIndex) {
+std::unique_ptr<BHNode> BHTree::InsertNucleon(std::unique_ptr<BHNode> node, const G4ThreeVector& cords, int pIndex) {
```

Comment:
Change of ownership is unnecessary here, change signature to `void foo(const unique_ptr&, ...)`